File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    void f(int t, vector<int>& v, int s, int sum, vector<int>& tmp, vector<vector<int>>* res){
        if(sum > t) return;
        if(sum == t){
            res->push_back(tmp);
            return;
        }
        for(int i = s; i < v.size(); ++i){
            tmp.push_back(v[i]);
            f(t, v, i, sum + v[i], tmp, res);
            tmp.pop_back();
        }
    }

    vector<vector<int>> combinationSum(vector<int>& v, int t) {
        vector<int> tmp;
        vector<vector<int>> ans;
        sort(v.begin(), v.end());
        f(t, v, 0, 0, tmp, &ans);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.erl
-spec combination_sum(Candidates :: [integer()], Target :: integer()) -> [[integer()]].
combination_sum(Candidates, Target) ->
    bcktrk(Candidates, Target, 0, 0, []).

bcktrk(_, Target, _, Sum, _) when Target < Sum -> [];
bcktrk(_, Target, _, Target, Tmp) -> [Tmp];
bcktrk(Candidates, Target, S, Sum, Tmp) ->
    element(2, lists:foldl(
        fun(X, {_S, Res}) ->
            {_S + 1, Res ++ bcktrk(Candidates, Target, _S, Sum + X, Tmp ++ [X])}
        end, {S, []}, element(2, lists:split(S, Candidates))
    )).